api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+//
+// Listeners that specify identical Hostname fields must not be
+// treated as collapsible. Put another way, each Hostname field in
+// a collapsible group must be unique in order to avoid ambiguity.
+//
+// If the GatewayClass collapses Listeners, the host name provided
+// in the incoming client request must be matched to a Listener to
+// find the correct set of Routes. The incoming host name must be matched
+// using the Hostname field for each listener in order of most to least
+// specific. That is, "Exact" matches must be processed before
+// "Domain" matches, which must be processed before Listeners that
+// omit the Hostname field.
+//
+// If the Gateway specifies multiple Listeners that have the same
+// Port value but are not collapsible, the GatewayClass must raise
+// a "PortConflict" condition on the Gateway.
 type Listener struct {
-	// Name is the listener's name and should be specified as an
-	// RFC 1035 DNS_LABEL [1]:
+	// Hostname specifies the virtual host name for protocol types
+	// that define this concept.
 	//
-	// [1] https://tools.ietf.org/html/rfc1035
+	// Incoming requests are matched against Hostname before processing
+	// HTTPRoute rules. For example, if the HTTP request header
+	// contains "Host: foo.example.com", an Listener with Hostname value
+	// "foo.example.com" will be an "Exact" match, but a Listener with
+	// Hostname values "example.com" or "bar.example.com" will not
+	// match. However these host names would match against a "Domain"
+	// match type with a value of "example.com".
 	//
-	// Each listener of a Gateway must have a unique name. Name is used
-	// for associating a listener in Gateway status.
+	// If Hostname is unspecified, the Gateway routes all traffic to
+	// the specified routes. This is equivalent to matching all host
+	// and domain names.
 	//
-	// Support: Core
+	// A Hostname match should be supplied for "HTTP" and "HTTPS"
+	// protocols, and may be supplied for "TLS" protocol. If no
+	// match is supplied, then this Listener must accept all client
+	// connections on the specified Port.
 	//
-	// +required
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
-	// Address requested for this listener. This is optional and behavior
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// As Listener that omits this field may be collapsed with other
+	// Listeners on the same port and protocol. There may be only
+	// one such Listener per Port. If more that one is configured,
+	// the GatewayClass must raise a "NameConflict" condition.
 	//
-	// Support:
+	// Support: Core
 	//
 	// +optional
-	Address *ListenerAddress `json:"address,omitempty" protobuf:"bytes,2,opt,name=address"`
-	// Port is a list of ports associated with the Address.
+	Hostname *HostnameMatch `json:"hostname,omitempty" protobuf:"bytes,1,opt,name=hostname"`
```

Comment:
Thinking more about this, why is this not an array instead?
Use case:
I want to control `*example.com` and `*.new.example.com` using a single Gateway.

Reply:
Reading through the examples I see how this would work. I'll post my comment in a comment on the PR itself rather than starting the conversation here.